File: host/main/scan.cpp
#include "scan.h"

FileMetadata Audio_Func(const fs::path& filePath) {
    FileMetadata meta;
    meta.path = filePath.string();
    meta.extension = filePath.extension().string();
    meta.fileSize = fs::file_size(filePath);
    
    static const std::vector<std::string> audioExts = {
        ".mp3", ".flac", ".ogg", ".wav", ".m4a", ".aac", ".wma"
    };
    
    bool isAudioFile = std::find(audioExts.begin(), audioExts.end(), meta.extension) != audioExts.end();
    
    if (isAudioFile) {
        TagLib::FileRef f(filePath.c_str());
        if (!f.isNull() && f.tag()) {
            TagLib::Tag* tag = f.tag();
            meta.title = tag->title().toCString(true);
            meta.artist = tag->artist().toCString(true);
            meta.album = tag->album().toCString(true);
            meta.year = tag->year();
        }
        meta.duration = ""; // Audio files could have duration extracted with additional libraries
    }
    
    return meta;
}

FileMetadata Video_Func(const fs::path& filePath) {
    FileMetadata meta;
    meta.path = filePath.string();
    meta.extension = filePath.extension().string();
    meta.fileSize = fs::file_size(filePath);
    
    static const std::vector<std::string> videoExts = {
        ".mp4", ".avi", ".mkv", ".mov", ".wmv", ".flv", ".webm", 
        ".m4v", ".mpg", ".mpeg", ".3gp", ".ogv", ".ts", ".m2ts"
    };
    
    bool isVideoFile = std::find(videoExts.begin(), videoExts.end(), meta.extension) != videoExts.end();
    
    if (isVideoFile) {
        // TODO: Finish this up
        meta.title = filePath.stem().string();
        meta.artist = "";
        meta.album = "";
        meta.year = 0;
        meta.duration = "3:32"; //<- I have no idea how to do this so I'm using examples
     }
    
    return meta;
}

// Standalone main function for scan target
#ifndef METEOR_MAIN_APP
int main() {
    FileScanner scanner;
    
    std::string directoryPath = ".";
    std::cout << "Enter directory path to scan (default: current directory): ";
    std::getline(std::cin, directoryPath);
    
    if (directoryPath.empty()) {
        directoryPath = ".";
    }
    
    std::cout << "Scanning directory: " << directoryPath << std::endl;
    scanner.scanDirectory(directoryPath);
    
    // Write metadata to file
    std::string outputFilename = "file_metadata_report.txt";
    scanner.writeMetadataToFile(outputFilename);
    
    // Also display to console
    scanner.printMetadata();
    
    return 0;
}
#endif

void FileScanner::scanDirectory(const std::string& directoryPath) {
    indexedFiles.clear();
    
    if (!fs::exists(directoryPath) || !fs::is_directory(directoryPath)) {
        std::cerr << "Error: Directory does not exist: " << directoryPath << std::endl;
        return;
    }
    
    try {
        for (const auto& entry : fs::recursive_directory_iterator(directoryPath)) {
            if (entry.is_regular_file()) {
                FileMetadata meta;
                std::string extension = entry.path().extension().string();
                
                // Transform extension to lowercase for comparison
                std::transform(extension.begin(), extension.end(), extension.begin(), ::tolower);
                
                // Check if it's an audio file
                static const std::vector<std::string> audioExts = {
                    ".mp3", ".flac", ".ogg", ".wav", ".m4a", ".aac", ".wma"
                };
                static const std::vector<std::string> videoExts = {
                    ".mp4", ".avi", ".mkv", ".mov", ".wmv", ".flv", ".webm", 
                    ".m4v", ".mpg", ".mpeg", ".3gp", ".ogv", ".ts", ".m2ts"
                };
                
                bool isAudioFile = std::find(audioExts.begin(), audioExts.end(), extension) != audioExts.end();
                bool isVideoFile = std::find(videoExts.begin(), videoExts.end(), extension) != videoExts.end();
                
                if (isAudioFile) {
                    meta = Audio_Func(entry.path());
                } else if (isVideoFile) {
                    meta = Video_Func(entry.path());
                } else {
                    // For other files, create basic metadata
                    meta.path = entry.path().string();
                    meta.extension = entry.path().extension().string();
                    meta.fileSize = fs::file_size(entry.path());
                    meta.title = entry.path().stem().string();
                    meta.artist = "";
                    meta.album = "";
                    meta.year = 0;
                    meta.duration = "";
                }
                
                indexedFiles.push_back(meta);
            }
        }
    } catch (const fs::filesystem_error& e) {
        std::cerr << "Filesystem error: " << e.what() << std::endl;
    }
}
    
void FileScanner::writeMetadataToFile(const std::string& filename) const {
    std::ofstream outFile(filename);
    if (!outFile.is_open()) {
        std::cerr << "Error: Could not open file " << filename << " for writing." << std::endl;
        return;
    }
    
    outFile << "File Metadata Report\n";
    outFile << "===================\n\n";
    outFile << "Total files indexed: " << indexedFiles.size() << "\n\n";
    
    for (const auto& file : indexedFiles) {
        outFile << "File: " << file.path << "\n";
        outFile << "  Size: " << file.fileSize << " bytes\n";
        outFile << "  Extension: " << file.extension << "\n";
        outFile << "  Title: " << file.title << "\n";
        
        if (!file.artist.empty()) outFile << "  Artist: " << file.artist << "\n";
        if (!file.album.empty()) outFile << "  Album: " << file.album << "\n";
        if (file.year > 0) outFile << "  Year: " << file.year << "\n";
        
        outFile << "\n";
    }
    
    outFile.close();
    std::cout << "Metadata written to file: " << filename << std::endl;
}
    
const std::vector<FileMetadata>& FileScanner::getIndexedFiles() const {
    return indexedFiles;
}
    
void FileScanner::printMetadata() const {
    std::cout << "Indexed " << indexedFiles.size() << " files:\n" << std::endl;
    
    for (const auto& file : indexedFiles) {
        std::cout << "File: " << file.path << std::endl;
        std::cout << "  Size: " << file.fileSize << " bytes" << std::endl;
        std::cout << "  Extension: " << file.extension << std::endl;
        
        if (!file.title.empty() || !file.artist.empty() || !file.album.empty()) {
            std::cout << "  Metadata:" << std::endl;
            if (!file.title.empty()) std::cout << "    Title: " << file.title << std::endl;
            if (!file.artist.empty()) std::cout << "    Artist: " << file.artist << std::endl;
            if (!file.album.empty()) std::cout << "    Album: " << file.album << std::endl;
            if (file.year > 0) std::cout << "    Year: " << file.year << std::endl;
        }
        std::cout << std::endl;
    }
}
    
std::vector<FileMetadata> FileScanner::searchByArtist(const std::string& artist) const {
    std::vector<FileMetadata> results;
    for (const auto& file : indexedFiles) {
        if (file.artist == artist) {
            results.push_back(file);
        }
    }
    return results;
}
    
std::vector<FileMetadata> FileScanner::searchByAlbum(const std::string& album) const {
    std::vector<FileMetadata> results;
    for (const auto& file : indexedFiles) {
        if (file.album == album) {
            results.push_back(file);
        }
    }
    return results;
}

File: host/scan.py
"""
host/scan.py — File metadata scanner (Python rewrite of host/main/scan.cpp)
-----------------------------------------------------------------------------
Recursively walks a directory, extracts metadata from audio files via
mutagen (if installed), and produces a text report.

Usage
-----
    from host.scan import FileScanner

    scanner = FileScanner()
    scanner.scan_directory("/path/to/music")
    scanner.write_metadata_to_file("report.txt")
    for f in scanner.indexed_files:
        print(f["path"], f["title"])

Or run directly:
    python3 host/scan.py
"""

import os
import sys

_AUDIO_EXTS = {".mp3", ".flac", ".ogg", ".wav", ".m4a", ".aac", ".wma"}
_VIDEO_EXTS = {".mp4", ".avi", ".mkv", ".mov", ".wmv", ".flv", ".webm",
               ".m4v", ".mpg", ".mpeg", ".3gp", ".ogv", ".ts", ".m2ts"}

# ── Optional mutagen for audio tag reading ────────────────────────────────────
try:
    import mutagen
    _MUTAGEN = True
except ImportError:
    _MUTAGEN = False


# ── Internal helpers ──────────────────────────────────────────────────────────

def _audio_metadata(path: str) -> dict:
    """
    Extract audio metadata from a file using mutagen when available.
    Falls back to filename stem as title when mutagen is not installed.
    """
    stem = os.path.splitext(os.path.basename(path))[0]
    meta = {
        "path":      path,
        "extension": os.path.splitext(path)[1].lower(),
        "file_size": os.path.getsize(path),
        "title":     stem,
        "artist":    "",
        "album":     "",
        "year":      0,
        "duration":  "",
    }

    if _MUTAGEN:
        try:
            from mutagen import File as MFile
            audio = MFile(path, easy=True)
            if audio and audio.tags:
                meta["title"]  = (audio.tags.get("title",  [stem])[0]  or stem)
                meta["artist"] = (audio.tags.get("artist", [""])[0]    or "")
                meta["album"]  = (audio.tags.get("album",  [""])[0]    or "")
                try:
                    meta["year"] = int(audio.tags.get("date", ["0"])[0] or 0)
                except (ValueError, TypeError):
                    meta["year"] = 0
                if hasattr(audio, "info") and hasattr(audio.info, "length"):
                    secs = int(audio.info.length)
                    meta["duration"] = f"{secs // 60}:{secs % 60:02d}"
        except Exception:
            pass  # malformed file — keep defaults

    return meta


def _video_metadata(path: str) -> dict:
    """
    Build basic metadata for a video file.
    Full duration extraction would require an external library (e.g. ffprobe).
    """
    stem = os.path.splitext(os.path.basename(path))[0]
    return {
        "path":      path,
        "extension": os.path.splitext(path)[1].lower(),
        "file_size": os.path.getsize(path),
        "title":     stem,
        "artist":    "",
        "album":     "",
        "year":      0,
        "duration":  "",  # TODO: use ffprobe / mediainfo to get real duration
    }


def _basic_metadata(path: str) -> dict:
    """Minimal metadata for non-media files."""
    stem = os.path.splitext(os.path.basename(path))[0]
    return {
        "path":      path,
        "extension": os.path.splitext(path)[1].lower(),
        "file_size": os.path.getsize(path),
        "title":     stem,
        "artist":    "",
        "album":     "",
        "year":      0,
        "duration":  "",
    }


# ── FileScanner class (mirrors C++ FileScanner) ───────────────────────────────

class FileScanner:
    """
    Recursively scans a directory and collects file metadata.

    Attributes
    ----------
    indexed_files : list[dict]
        All discovered files as dicts with keys:
        path, extension, file_size, title, artist, album, year, duration
    """

    def __init__(self):
        self.indexed_files: list[dict] = []

    def scan_directory(self, directory: str) -> None:
        """
        Walk *directory* recursively, building the indexed_files list.
        The list is cleared on every call.
        """
        self.indexed_files = []

        if not os.path.isdir(directory):
            print(f"Error: Directory does not exist: {directory}", file=sys.stderr)
            return

        for root, _dirs, files in os.walk(directory):
            for name in files:
                path = os.path.join(root, name)
                ext  = os.path.splitext(name)[1].lower()

                try:
                    if ext in _AUDIO_EXTS:
                        meta = _audio_metadata(path)
                    elif ext in _VIDEO_EXTS:
                        meta = _video_metadata(path)
                    else:
                        meta = _basic_metadata(path)

                    self.indexed_files.append(meta)
                except OSError:
                    pass  # skip files we can't read

    def write_metadata_to_file(self, filename: str) -> None:
        """Write a human-readable report to *filename*."""
        try:
            with open(filename, "w", encoding="utf-8") as f:
                f.write("File Metadata Report\n")
                f.write("===================\n\n")
                f.write(f"Total files indexed: {len(self.indexed_files)}\n\n")

                for meta in self.indexed_files:
                    f.write(f"File: {meta['path']}\n")
                    f.write(f"  Size: {meta['file_size']} bytes\n")
                    f.write(f"  Extension: {meta['extension']}\n")
                    f.write(f"  Title: {meta['title']}\n")
                    if meta.get("artist"):
                        f.write(f"  Artist: {meta['artist']}\n")
                    if meta.get("album"):
                        f.write(f"  Album: {meta['album']}\n")
                    if meta.get("year", 0) > 0:
                        f.write(f"  Year: {meta['year']}\n")
                    f.write("\n")

            print(f"Metadata written to file: {filename}")
        except OSError as exc:
            print(f"Error: Could not write to {filename}: {exc}", file=sys.stderr)

    def print_metadata(self) -> None:
        """Print a summary of all indexed files to stdout."""
        print(f"Indexed {len(self.indexed_files)} files:\n")
        for meta in self.indexed_files:
            print(f"File: {meta['path']}")
            print(f"  Size: {meta['file_size']} bytes")
            print(f"  Extension: {meta['extension']}")
            if meta.get("title") or meta.get("artist") or meta.get("album"):
                print("  Metadata:")
                if meta.get("title"):  print(f"    Title: {meta['title']}")
                if meta.get("artist"): print(f"    Artist: {meta['artist']}")
                if meta.get("album"):  print(f"    Album: {meta['album']}")
                if meta.get("year", 0) > 0: print(f"    Year: {meta['year']}")
            print()

    def search_by_artist(self, artist: str) -> list[dict]:
        """Return all files whose artist field matches *artist* exactly."""
        return [f for f in self.indexed_files if f.get("artist") == artist]

    def search_by_album(self, album: str) -> list[dict]:
        """Return all files whose album field matches *album* exactly."""
        return [f for f in self.indexed_files if f.get("album") == album]


# ── Direct run ────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    directory = input("Enter directory path to scan (default: current directory): ").strip()
    if not directory:
        directory = "."

    print(f"Scanning directory: {directory}")
    scanner = FileScanner()
    scanner.scan_directory(directory)
    scanner.write_metadata_to_file("file_metadata_report.txt")
    scanner.print_metadata()
